src/replication/replication_manager.cpp

```diff
@@ -1,81 +1,63 @@
-      last_heartbeat_(0) {}
-
-const char *ReplicateBufferMessage::key_buf_id = "buf_id";
-const char *ReplicateBufferMessage::key_content = "content";
-
-ReplicateBufferMessage ReplicateBufferMessage::FromMessage(const messenger::ZmqMessage &msg) {
-  // TODO(WAN): Sanity-check the received message.
-  common::json message = nlohmann::json::parse(msg.GetMessage());
-  uint64_t source_callback_id = msg.GetSourceCallbackId();
-  uint64_t buffer_id = message.at(key_buf_id);
-  std::string contents = nlohmann::json::from_cbor(message[key_content].get<std::vector<uint8_t>>());
-  return ReplicateBufferMessage(buffer_id, std::move(contents), source_callback_id);
-}
-
-common::json ReplicateBufferMessage::ToJson() {
-  common::json json;
-  json[key_buf_id] = buffer_id_;
-  json[key_content] = nlohmann::json::to_cbor(contents_);
-  // TODO(WAN): Add a size and checksum to message.
-  return json;
-}
-
 ReplicationManager::ReplicationManager(
     common::ManagedPointer<messenger::Messenger> messenger, const std::string &network_identity, uint16_t port,
     const std::string &replication_hosts_path,
     common::ManagedPointer<common::ConcurrentBlockingQueue<storage::BufferedLogWriter *>> empty_buffer_queue)
-    : messenger_(messenger), identity_(network_identity), port_(port), empty_buffer_queue_(empty_buffer_queue) {
-  auto listen_destination = messenger::ConnectionDestination::MakeTCP("", "127.0.0.1", port);
-  messenger_->ListenForConnection(listen_destination, network_identity,
-                                  [this](common::ManagedPointer<messenger::Messenger> messenger,
-                                         const messenger::ZmqMessage &msg) { EventLoop(messenger, msg); });
-  BuildReplicaList(replication_hosts_path);
+    : empty_buffer_queue_(empty_buffer_queue), messenger_(messenger), identity_(network_identity), port_(port) {
+  // Start listening on the given replication port.
+  messenger::ConnectionDestination listen_destination =
+      messenger::ConnectionDestination::MakeTCP("", "127.0.0.1", port);
+  messenger_->ListenForConnection(
+      listen_destination, network_identity,
+      [this](common::ManagedPointer<messenger::Messenger> messenger, const messenger::ZmqMessage &msg) {
+        auto json = nlohmann::json::parse(msg.GetMessage());
+        auto replication_msg = BaseReplicationMessage::ParseFromJson(json);
+        EventLoop(messenger, msg, common::ManagedPointer(replication_msg));
+      });
+  // Connect to all of the other nodes.
+  BuildReplicationNetwork(replication_hosts_path);
 }
 
-const char *ReplicationManager::key_message_type = "message_type";
-
 ReplicationManager::~ReplicationManager() = default;
 
-void ReplicationManager::EnableReplication() {
-  REPLICATION_LOG_TRACE(fmt::format("[PID={}] Replication enabled.", ::getpid()));
-  replication_enabled_ = true;
+msg_id_t ReplicationManager::GetNextMessageId() {
+  msg_id_t next_msg_id = next_msg_id_++;
+  if (next_msg_id_.load() == INVALID_MSG_ID) {
```

Comment:
I don't think this if statement is atomic (see my previous comments for more details).

Reply:
I think you're correct. I'm actually not sure how you would fix it off the top of my head without making it more expensive (ie a latch); wraparound is expected to be very rare / never. Currently, a "gap" of missing message IDs at the front doesn't matter for any application, and I think message IDs are ultimately dumb identifiers / there is nothing relying on receiving consecutive message IDs. So I added a TODO but that's about it, if you're OK with that then mark as resolve otherwise hmm we should discuss it further.

Reply:
actually, I might just be a monkey.

Reply:
As far as I can tell, this will ensure that every message id is unique, no message ids are skipped, and no message id will be equal to `INVALID_MSG_ID`. It could end up assigning "out of order" message ids, in the sense that if thread1 requests a message id before thread2 then thread1 *might* get a higher message id than thread 2. But I don't think that's a problem here.

Also, If `++` is more efficient than `compare_exchange_strong`, it might be worth switching the implementation for `Messenger::GetNextSendCallbackId` to this approach. Though it's probably not too important.

Reply:
1. yea, that wouldn't be a problem.
2. I changed `GetNextSendCallbackId`, but that one is slightly more complicated (while instead of if) because it actually has a range of values reserved.